### PR TITLE
Support receiving rules events

### PR DESCRIPTION
**Pull request by aounhaider1**
1) Receive rule events and update local repo accordingly.
2) Rely more on rule repository (local) than rule service (remote).
3) Added Status info of rule.
fixes https://github.com/eclipse/smarthome/issues/2049
Signed-off-by: Aoun Bukhari <bukhari@itemis.de>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
@maggu2810 I leave this to you to test and merge.

**Comment by maggu2810** [comment]
* I merged this PR locally
* I opened the Rules list in two different browsers
* I added a new rule in browser A
* if I press an the + button in browser B (so the contect menu "new rule" "rule from template" is shown), the list is repainted and I see the newly created rule


So it seems the event is correctly handled and the new rule is added internally.
I would expect that the new rule is also added to the currently shown list (e.g. at the end to not change the order). Why do we need an user interaction to show the new rule?

File: extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js [hunk 1]
```diff
@@ -161,8 +161,48 @@ angular.module('PaperUI.services.repositories', []).factory('bindingRepository',
...
+
+    eventService.onEvent('smarthome/rules/*/added', function(topic, rule) {
+        repository.add(rule);
+    });
+
+    eventService.onEvent('smarthome/rules/*/removed', function(topic, removedRule) {
+        var existing = repository.find(function(rule) {
+            return rule.uid === removedRule.uid;
+        });
+        repository.remove(existing);
```
Comment (maggu2810) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js]: Shouldn't we use `if (existing)` similar to the update event here, too?

File: extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js [hunk 2]
```diff
@@ -161,8 +161,48 @@ angular.module('PaperUI.services.repositories', []).factory('bindingRepository',
...
+    });
+
+    eventService.onEvent('smarthome/rules/*/state', function(topic, rule) {
+        var existing = repository.find(function(rule) {
+            return rule.uid === topic.split('/')[2];
+        });
+        existing.status = existing.status ? existing.status : {};
```
Comment (maggu2810) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js]: `if (existing)`?

File: extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js [hunk 3]
```diff
@@ -161,8 +161,48 @@ angular.module('PaperUI.services.repositories', []).factory('bindingRepository',
...
+        existing.status.status = rule.status;
+        existing.status.statusDetails = rule.status;
```
Comment (maggu2810) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js]: Why set we status.status to rule.status and stats.statusDetails also to rule.status? Shouldn't the details point to the details?

File: extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js (hunk 1 above)
Comment (aounhaider1) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js]: Not necessarily, I would expect remove function in repository to take care of that.

**Comment by aounhaider1** [comment]
The new rule list is sorted so every time a new rule is added angular insert it on the right position. I think it makes sense to maintain the order however, we can have the same highlight feature as we have in things to make it easier to identify the newly created rules. What do you say?

```Why do we need an user interaction to show the new rule?```
I did not understand what you mean here?

**Comment by maggu2810** [comment]
> (1) I merged this PR locally
> (2) I opened the Rules list in two different browsers
> (3) I added a new rule in browser A
> (4) if I press an the + button in browser B (so the contect menu "new rule" "rule from template" is shown), the list is repainted and I see the newly created rule

See number (4), if I don't press e.g. on the + button the new rule is notshown in browser B. I need to interact with the UI so it gets refreshed.

**Comment by aounhaider1** [comment]
The scope should refresh now when new rule events are received.

File: extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js (hunk 3 above)
Comment (aounhaider1) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js]: You are right it should be rule.statusDetails.

Comment (aounhaider1) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js]: Changed it!

File: extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js (hunk 1 above)
Comment (aounhaider1) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/services.repositories.js]: Added if found check on remove function.

**Comment by maggu2810** [comment]
Looks good now, thanks

**Comment by maggu2810** [comment]
Have you finished your changes?

**Comment by aounhaider1** [comment]
Yes.

**Comment by maggu2810** [comment]
Thanks

**Pull request merged**